src/lib/cache_lib/impl.ml

```diff
@@ -13,49 +12,54 @@ module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
 
     val logger : _ t -> Logger.t
 
-    val remove : 'elt t -> 'elt -> unit Or_error.t
+    val remove : 'elt t -> 'elt -> unit
   end = struct
-    type 'a t = {name: string; set: 'a Hash_set.t; logger: Logger.t}
+    type 'a t =
+      {name: string; set: ('a, 'a Intf.final_state) Hashtbl.t; logger: Logger.t}
 
     let name {name; _} = name
 
     let logger {logger; _} = logger
 
     let create (type elt) ~name ~logger
-        (module Elt : Hash_set.Elt_plain with type t = elt) : elt t =
-      let set =
-        Hash_set.create ~growth_allowed:true ?size:None (module Elt) ()
-      in
+        (module Elt : Hashtbl.Key_plain with type t = elt) : elt t =
+      let set = Hashtbl.create ~growth_allowed:true ?size:None (module Elt) in
       let logger = Logger.extend logger [("cache", `String name)] in
       {name; set; logger}
 
+    let final_state t x = Hashtbl.find t.set x
+
     let register t x =
-      let%map () = Hash_set.strict_add t.set x in
-      Cached.create t x
+      let final_state = Ivar.create () in
+      Hashtbl.add_exn t.set ~key:x ~data:final_state ;
```

Comment:
It would be nice to maintain the old interface of returning an `Or_error.t` instead of throwing exceptions.